server/controller/src/main/java/ai/starwhale/mlops/datastore/ColumnType.java

```diff
@@ -89,102 +101,168 @@ public static int compare(ColumnType type1, Object value1, ColumnType type2, Obj
+                return ((Short) value1).compareTo((Short) value2);
+            } else if (type1 == ColumnTypeScalar.INT32) {
+                return ((Integer) value1).compareTo((Integer) value2);
+            } else if (type1 == ColumnTypeScalar.INT64) {
+                return ((Long) value1).compareTo((Long) value2);
+            } else if (type1 == ColumnTypeScalar.FLOAT32) {
+                return ((Float) value1).compareTo((Float) value2);
+            } else if (type1 == ColumnTypeScalar.FLOAT64) {
+                return ((Double) value1).compareTo((Double) value2);
+            } else if (type1 == ColumnTypeScalar.STRING) {
+                return ((String) value1).compareTo((String) value2);
+            } else if (type1 == ColumnTypeScalar.BYTES) {
+                return ((ByteBuffer) value1).compareTo((ByteBuffer) value2);
+            }
+        } else {
+            var category1 = type1.getCategory();
+            var category2 = type2.getCategory();
+            if (category1.equals(category2)) {
+                if (category1.equals(ColumnTypeScalar.INT32.getCategory())) {
+                    return Long.compare(((Number) value1).longValue(), ((Number) value2).longValue());
+                } else if (category1.equals(ColumnTypeScalar.FLOAT32.getCategory())) {
+                    return Double.compare(((Number) value1).doubleValue(), ((Number) value2).doubleValue());
                 }
             } else {
-                var category1 = ((ColumnTypeScalar) type1).getCategory();
-                var category2 = ((ColumnTypeScalar) type2).getCategory();
-                if (category1.equals(category2)) {
-                    if (category1.equals(ColumnTypeScalar.INT32.getCategory())) {
-                        return Long.compare(((Number) value1).longValue(), ((Number) value2).longValue());
-                    } else if (category1.equals(ColumnTypeScalar.FLOAT32.getCategory())) {
-                        return Double.compare(((Number) value1).doubleValue(), ((Number) value2).doubleValue());
-                    }
-                } else {
-                    if (category1.equals(ColumnTypeScalar.INT32.getCategory())
-                            && category2.equals(ColumnTypeScalar.FLOAT32.getCategory())) {
-                        if (type1 == ColumnTypeScalar.INT64) {
-                            long v1 = (Long) value1;
-                            if (v1 > (1L << 53)) {
-                                return BigDecimal.valueOf(v1)
-                                        .compareTo(BigDecimal.valueOf(((Number) value2).doubleValue()));
-                            }
+                if (category1.equals(ColumnTypeScalar.INT32.getCategory())
+                        && category2.equals(ColumnTypeScalar.FLOAT32.getCategory())) {
+                    if (type1 == ColumnTypeScalar.INT64) {
+                        long v1 = (Long) value1;
+                        if (v1 > (1L << 53)) {
```

Comment:
so, considering the precision of js, can't we store more than 53 digits? shouldn't such precision issues be considered in server?

Reply:
Here is the logic of comparing an int64 with a float. Integers greater than 1^53 can not be converted to Double without precision loss. That's why here is a if statement. BigDecimal is used if the value is greater than 1^53, or Double otherwise.